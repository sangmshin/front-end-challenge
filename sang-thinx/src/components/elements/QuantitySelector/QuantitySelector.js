import React, { useState } from 'react';
import { number } from 'prop-types';
import { Button } from 'components/elements';
import styles from './QuantitySelector.module.scss';

export const QuantitySelector = ({ max }) => {
  const [quantity, setQuantity] = useState(1);
  const increaseQuantity = () => setQuantity(quantity + 1);
  const decreaseQuantity = () => setQuantity(quantity - 1);
  const disabledDecreaseButton = quantity === 1;
  const disabledIncreaseButton = quantity === max;

  return (
    <div className={styles.root}>
      <Button
        ariaLabel='decrease selected quantity'
        className={styles['decrease-button']}
        disabled={disabledDecreaseButton}
        onClick={decreaseQuantity}
        role='presentation'
        type='button'
      >
        -
      </Button>
      <span className={styles['quantity-text-wrapper']} role='presentation'>
        <span className={styles['quantity-text']}>
          quantity
        </span>
        <input
          aria-label='quantity'
          className={styles['quantity-number']}
          maxLength={max}
          name='quantity'
          readOnly
          step='1'
          value={`(${quantity})`}
        />
      </span>
      <Button
        ariaLabel='increase selected quantity'
        className={styles['increase-button']}
        disabled={disabledIncreaseButton}
        onClick={increaseQuantity}
        role='presentation'
        type='button'
      >
        +
      </Button>
    </div>
  )
}

QuantitySelector.propTypes = {
  max: number
}

QuantitySelector.defaultProps = {
  max: null
}

export default QuantitySelector;