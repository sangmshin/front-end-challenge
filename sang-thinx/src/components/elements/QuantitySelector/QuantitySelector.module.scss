@import 'variables';

.root {
  display: flex;
  height: 30px;
  border: 1px solid black;
  margin-top: 20px;

  .decrease-button {
    border-right: 1px solid black;
  }

  .increase-button {
    border-left: 1px solid black;
  }

  .decrease-button, .increase-button {
    width: 40px;
    font-size: 1.25rem;
  }

  .quantity-text-wrapper {
    flex: 1;
    padding: 3px 10px;
    text-align: center;

    .quantity-text {
      font-size: 0.875rem;
      padding-left: 30px;
    }

    .quantity-number {
      width: 40px;
      background-color: transparent;
      font-size: 0.875rem;
      border: none;
    }
  }
}

@media (min-width: $screen-md) {
  .root .quantity-text-wrapper .quantity-text,
  .root .quantity-text-wrapper .quantity-number {
      font-size: 1rem;
  }
}