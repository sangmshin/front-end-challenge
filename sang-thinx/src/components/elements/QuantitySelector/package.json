{
  "name": "QuantitySelector",
  "main": "./QuantitySelector.js"
}
