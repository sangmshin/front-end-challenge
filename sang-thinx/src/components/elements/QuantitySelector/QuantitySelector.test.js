import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { QuantitySelector } from './QuantitySelector';

const MAX_QUANTITY = 99;

describe('Component :: QuantitySelector', () => {
  
  describe('#render', () => {
    const component = shallow(
      <QuantitySelector />
    );
    it('should match snapshot', () => {
      expect(toJson(component)).toMatchSnapshot();
    });
  });

  describe('#max prop', () => {
    const component = shallow(
      <QuantitySelector max={MAX_QUANTITY} />
    );
    it('should have input element that has max-length attribute', () => {
      expect(component.find('input').prop('maxLength')).toBe(MAX_QUANTITY);
    });
  });
});
