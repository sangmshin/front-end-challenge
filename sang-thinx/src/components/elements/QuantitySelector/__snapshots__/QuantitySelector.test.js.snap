// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component :: QuantitySelector #render should match snapshot 1`] = `
<div
  className="root"
>
  <ForwardRef
    accessibleText=""
    ariaLabel="decrease selected quantity"
    className="decrease-button"
    disabled={true}
    onClick={[Function]}
    role="presentation"
    style={null}
    type="button"
  >
    -
  </ForwardRef>
  <span
    className="quantity-text-wrapper"
    role="presentation"
  >
    <span
      className="quantity-text"
    >
      quantity
    </span>
    <span>
       (
    </span>
    <input
      aria-label="quantity"
      className="quantity-number"
      maxLength={null}
      name="quantity"
      readOnly={true}
      step="1"
      type="number"
      value="1"
    />
    <span
      className="closing-parenthesis"
    >
      )
    </span>
  </span>
  <ForwardRef
    accessibleText=""
    ariaLabel="increase selected quantity"
    className="increase-button"
    disabled={false}
    onClick={[Function]}
    role="presentation"
    style={null}
    type="button"
  >
    +
  </ForwardRef>
</div>
`;
